mrmustard/lab_dev/states/base.py

```diff
@@ -796,6 +899,51 @@ def dm(self) -> DM:
         dm = self @ self.adjoint
         return DM._from_attributes(self.name, dm.representation, dm.wires)
 
+    def expectation(self, operator: CircuitComponent):
+        r"""
+        The expectation value of an operator calculated over this Ket.
+
+        Given the operator `O`, this function returns :math:`Tr\big(|\psi\rangle\langle\psi| O)`\,
+        where :math:`|\psi\rangle` is the vector representing this state.
+
+        The ``operator`` is expected to be a component with ket-like wires (i.e., output wires on
+        the ket side), density matrix-like wires (output wires on both ket and bra sides), or
+        unitary-like wires (input and output wires on the ket side).
+
+        Args:
+            operator: A ket-like, density-matrix like, or unitary-like circuit component.
+
+        Raise:
+            ValueError: If ``operator`` is not a ket-like, density-matrix like, or unitary-like
+                component.
+            ValueError: If ``operator`` is defined over a set of modes that is not a subset of the
+                modes of this state.
+        """
+        op_type, msg = _validate_operator(operator)
+        if op_type is OperatorType.INVALID_TYPE:
+            raise ValueError(msg)
+
+        if not operator.wires.modes.issubset(self.wires.modes):
+            msg = f"Expected an operator defined on a subset of modes `{self.modes}`, "
+            msg += f"found one defined on `{operator.modes}.`"
+            raise ValueError(msg)
+
+        leftover_modes = self.wires.modes - operator.wires.modes
+        if op_type is OperatorType.KET_LIKE:
+            result = self @ operator.dual
+            if leftover_modes:
+                result >>= TraceOut(leftover_modes)
+            result @= result.dual
```

Comment:
this should not happen if you trace out, because the result of trace out is already a dm

Reply:
ooh well spotted!!